### PR TITLE
DX-633: Add Vector Definitions

### PR DESCRIPTION
**Pull request by fahreddinozcan**

### CODE REVIEW

File: integrationtesting/upstash_vector_index_test.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	"strconv"
+	"testing"
+
+	"github.com/gruntwork-io/terratest/modules/terraform"
+)
+
+var vectorIndexName, vectorIndexType, vectorIndexRegion, vectorIndexSimilarityFunction, vectorIndexDimensionCount string
+
+func TestUpstashVectorIndexMAIN(t *testing.T) {
+	//t.Parallel()
```
Comment (burak-upstash) [inline integrationtesting/upstash_vector_index_test.go:14]: let's remove if not used

File: upstash/vector/index/resource.go
Comment (burak-upstash) [inline upstash/vector/index/resource.go:1]: Can rename this directory as upstash/vector/resource.go, and all the other ones? I don't think we will have any other resources under the name vector, so no need for subdirectory. Wdyt?

File: upstash/vector/index/types.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	Endpoint             string `json:"endpoint"`
+	Token                string `json:"token"`
+	ReadOnlyToken        string `json:"read_only_token"`
+	Type                 string `json:"type"`
+	Region               string `json:"region"`
+	CreationTime         int    `json:"creation_time"`
+	MaxVectorCount       int    `json:"max_vector_count"`
+	MaxDailyUpdates      int    `json:"max_daily_updates"`
+	MaxDailyQueries      int    `json:"max_daily_queries"`
+	MaxMonthlyBandwidth  int    `json:"max_monthly_bandwidth"`
```
Comment (burak-upstash) [inline upstash/vector/index/types.go:17]: this can be int64

File: upstash/vector/index/types.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	MaxWritesPerSecond   int    `json:"max_writes_per_second"`
+	MaxQueryPerSecond    int    `json:"max_query_per_second"`
+	MaxReadsPerRequest   int    `json:"max_reads_per_request"`
+	MaxWritesPerRequest  int    `json:"max_writes_per_request"`
+	MaxTotalMetadataSize int    `json:"max_total_metadata_size"`
+	ReservedPrice        int    `json:"reserved_price"`
```
Comment (burak-upstash) [inline upstash/vector/index/types.go:23]: this can be float64

File: upstash/vector/index/types.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	Type               string `json:"type"`
+}
+
+type SetPlanRequest struct {
+	TargetPlan string `json:"target_plan"`
+}
+
+type TransferIndexRequest struct {
+	TargetAccount string `json:"target_account"`
+}
```
Comment (burak-upstash) [inline upstash/vector/index/types.go:41]: is this used?

File: upstash/vector/index/crud.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+import (
+	"context"
+	"github.com/upstash/terraform-provider-upstash/upstash/utils"
+
+	"github.com/upstash/terraform-provider-upstash/upstash/client"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	_ "github.com/upstash/terraform-provider-upstash/upstash/client"
+	_ "github.com/upstash/terraform-provider-upstash/upstash/utils"
```
Comment (burak-upstash) [inline upstash/vector/index/crud.go:12]: can we remove the `_` before imports

File: upstash/vector/index/api_calls.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	_, err = c.SendPostRequest("/v2/vector/index/"+indexId+"/reset-password", nil, "Reset Password for Vector Index", false)
+
+	return err
+}
+
+func TransferIndex(c *client.UpstashClient, indexId string, team string) (err error) {
+	_, err = c.SendPostRequest("/v2/vector/index/"+indexId+"/transfer", team, "Transfer Vector Index to Another Team", false)
+
+	return err
+}
```
Comment (burak-upstash) [inline upstash/vector/index/api_calls.go:49]: Transferring for now is a bit risky, there is a chance that user won't be able to manage their tf state after transferring. I would personally want to  remove this functionality unless someone requested it... Wdyt?

File: upstash/vector/index/crud.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	mapping := map[string]interface{}{
+		"customer_id":         index.CustomerId,
+		"id":                  index.Id,
+		"name":                index.Name,
+		"similarity_function": index.SimilarityFunction,
+		"dimension_count":     index.DimensionCount,
+		"endpoint":            index.Endpoint,
+		"token":               index.Token,
+		"read_only_token":     index.ReadOnlyToken,
+		//"type":                    index.Type,
```
Comment (burak-upstash) [inline upstash/vector/index/crud.go:68]: let's remove if not used

File: upstash/vector/index/resource.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+			},
+			"creation_time": &schema.Schema{
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "The creation time of the vector index in UTC as unix timestamp.",
+			},
+		},
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
```
Comment (burak-upstash) [inline upstash/vector/index/resource.go:129]: Have you tried importing?

File: upstash/vector/index/crud.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+	err = data.Set("id", index.Id)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceIndexRead(ctx, data, m)
+
+}
+func resourceIndexRead(ctx context.Context, data *schema.ResourceData, m interface{}) diag.Diagnostics {
```
Comment (burak-upstash) [inline upstash/vector/index/crud.go:40]: More like a consistency thing, but can you add a newline where a function ends and new one starts? Also, some functions have newlines at the very end, and some don't. Maybe let's uniformize this :D

**Review by burak-upstash — COMMENTED** [review]
Good job, there are small stuff that maybe we should have a look at. Also, can you generate the docs via tfplugindocs? https://github.com/hashicorp/terraform-plugin-docs

File: upstash/vector/index/resource.go
Comment (fahreddinozcan) [reply upstash/vector/index/resource.go:1]: It's more for convention than direct folder structure. Vector and index are different in this context and we're managing  the index resources. This applies for the redis as well

File: upstash/vector/index/api_calls.go (hunk 6 above)
Comment (fahreddinozcan) [reply upstash/vector/index/api_calls.go:49]: Is there an update we need to do, to make the transfers possible?

**Review by burak-upstash — APPROVED** [review]
thanks, looks good

**Pull request merged**